torch/csrc/cuda/nccl.cpp

```diff
@@ -811,6 +811,47 @@ void recv(
 #endif
 }
 
+
+void gather(
+    const at::Tensor& inputs,
+    at::Tensor& outputs,
+    ncclComm_t _comm,
+    at::cuda::CUDAStream& stream,
+    int32_t root) {
+#ifdef USE_NCCL
+#if defined(NCCL_MAJOR) && (NCCL_MAJOR == 2) && (NCCL_MAJOR * 10 + NCCL_MINOR) >= 27
+  using namespace torch::cuda::nccl::detail;
+
+  auto comm = to_nccl_comm(_comm);
+  int numranks, cur_rank;
+  NCCL_CHECK(ncclCommCount(comm, &numranks));
+  NCCL_CHECK(ncclCommUserRank(comm, &cur_rank));
+
+  size_t count = inputs.numel();
+  auto type = to_nccl_data_type(inputs);
+  const auto* sendbuff = reinterpret_cast<char*>(inputs.data_ptr());
+
+  NCCL_CHECK(ncclGroupStart());
+  NCCL_CHECK(ncclSend(sendbuff, count, type, root, comm, stream));
+  if (cur_rank == root)
+  {
+    for (int r = 0; r < numranks; r++)
+    {
+      auto* recvbuff =  reinterpret_cast<char*>(outputs[r].data_ptr());
```

Comment:
nit: outputs[r] needs to create a new Tensor object (not necessarily a copy) and might have some minor overhead.. Was wondering if this can just be done using output.data_ptr().

Reply:
The `[]` operator has a bunch of overhead and creates a new Tensor object: https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/TensorOperators.h.html#_ZNK2at6TensorixEN3c106ScalarE. It might be slightly faster to just use `.data_ptr()` directly.

Reply:
Although, I feel this is probably fine for now since directly manipulating `data_ptr()` would involve getting strides and dtype etc and some pointer arithmetic.